[Refactor] JSON attributes on Models [#70]

chrisw000 opened the pull request:
I've added JsonAttributes to the model objects to allow serialization to enums and also allow .NET naming conventions on properties.

In some cases I've swapped from string to guid - eg make usage of OrderId consistent.

**Therefore, most model objects have breaking changes.**


RE changes for Bug #74 - it possibly reversed some correct guids back to string?

- Withdrawals.Model.Coinbase; coinbase docs say the CoinbaseAccountId is a GUID, its been changed back to string
- Withdrawals.Model.Withdrawal; paymentMethodId has been changed from GUID to string
- Deposits.Models.Deposit; paymentMethodId is still a GUID - so now inconsistent with the withdrawal service....  

From the docs - I think PaymentMethodId should be GUID, and CoinbaseAccountId should be GUID ?

Also, the Services.Payments.Models.PaymentMethod.Id is a GUID in anycase.

I've left the type changes as they were made in #74 (ie back to string) - let me know if you want to swap back to GUID again?

Also - same subject of GUID / string.... re #53 -> this applies to the following services; Accounts, Deposits, Fills, Orders -> I have the change ready to go if you want it checked in with this lot?

chrisw000 commented:
Hi - I've only just gone through some emails from earlier in the week.

I've just seen that you'd written to maybe do a pull request per Service, to make reviewing easier... the simplest way to review is to look at the tests - then you'll see which properties / types have been changed.

Reviewing a large changeset on git is hard work - but if you took a copy of the files and dumped them ontop of your working copy - you can see all the diffs in Visual Studio which is faster and more in context.

Let me know what you think?

dougdellolio commented:
great! going through these now.... may take me some time to verify things. hang tight

dougdellolio commented on GDAXSharp/Services/Fills/FillLiquidity.cs:
```diff
@@ -0,0 +1,12 @@
+        [EnumMember(Value = "T")]
+        Taker
+    }
+}
```
add nl

dougdellolio commented on GDAXSharp/Services/Orders/OrdersService.cs:
```diff
@@ -33,91 +32,86 @@ public class OrdersService : AbstractService
+            return await PlaceOrderAsync(order);
+        }
+
+        private async Task<OrderResponse> PlaceOrderAsync(Order order)
```
this is a nice refactor. 👍

dougdellolio commented on GDAXSharp/Services/Withdrawals/WithdrawalsService.cs:
```diff
@@ -45,16 +43,13 @@ public class WithdrawalsService : AbstractService
-            string coinbaseAccountId, 
-            decimal amount, 
-            Currency currency)
+        public async Task<CoinbaseResponse> WithdrawToCoinbaseAsync(string coinbaseAccountId, decimal amount, Currency currency)
```
please new line these as they were before

dougdellolio commented on GDAXSharp/Services/Withdrawals/Models/Withdrawal.cs:
```diff
@@ -1,11 +1,17 @@
+        public decimal Amount { get; set; }
 
-        public string currency { get; set; }
+        [JsonConverter(typeof(StringEnumConverter))]
```
need to add `[JsonProperty("currency")]` otherwise it won't serialize properly

dougdellolio commented on GDAXSharp/Services/Withdrawals/Models/Coinbase.cs:
```diff
@@ -1,11 +1,16 @@
     public class Coinbase
     {
-        public decimal amount { get; set; }
+        public decimal Amount { get; set; }
```
need to add `[JsonProperty("amount")]` here otherwise it won't serialize properly

dougdellolio commented on GDAXSharp/Services/Withdrawals/Models/Coinbase.cs:
```diff
@@ -1,11 +1,16 @@
 
-        public string currency { get; set; }
+        [JsonConverter(typeof(StringEnumConverter))]
+        public Shared.Currency Currency { get; set; }
```
need to add `[JsonProperty("currency")]` here otherwise it won't serialize properly

chrisw000 commented:
I've made the majority of the changes requested... comments below:


GDAXSharp/GDAXSharp.csproj.DotSettings
- This is the resharper solution settings file. I've deleted it, it might come back - add it to .gitignore?


In GDAXSharp/Services/CoinbaseAccounts/CoinbaseAccountType.cs
- I originally left this comment in, because the values are not found in the GDAX API documentation.. (its in the coinbase docs and so not instantly findable) I've removed anyway


In GDAXSharp/Services/Withdrawals/Models/Crypto.cs:  (and some others)

These all get serialized as lowercase automatically - so we don't need to tell it the lowercase name via [JsonProperty("x_y_z")].

**BUT!** I've found there is an out the box contract resolver for the GDAX snake case naming convention, meaning we don't need **any** of the [JsonProperty("x_y_z")] decorations 

var settings = new JsonSerializerSettings
{
    ContractResolver = new DefaultContractResolver
    {
        NamingStrategy = new SnakeCaseNamingStrategy()
    }
};

Assuming this works as expected, I'll remove all the [JsonProperty("x_y_z")] and get it to use the correct settings instead...

dougdellolio commented on GDAXSharp.Specs/Utilities/Extensions/ProductTypeExtensionsSpecs.cs:
```diff
@@ -7,56 +7,76 @@ namespace GDAXSharp.Specs.Utilities.Extensions
-
-        static string product_type_result;
+        static ProductType productId;
+        static Currency baseCurrency;
```
nl  between fields

dougdellolio commented on GDAXSharp/Services/Products/ProductsService.cs:
```diff
@@ -103,6 +103,8 @@ public class ProductsService : AbstractService
 
             do
             {
+                if (batchEnd == null) break;
```
please include brackets here:

````
 if (batchEnd == null) {
   break;
 }
````

dougdellolio commented on GDAXSharp/Services/Products/ProductsService.cs:
```diff
@@ -40,57 +40,57 @@ public class ProductsService : AbstractService
             return productsResponse;
         }
 
         public async Task<ProductsOrderBookResponse> GetProductOrderBookAsync(
```
please remove the unused using statement at the top

dougdellolio commented on GDAXSharp/Services/Orders/OrdersService.cs:
```diff
@@ -33,95 +32,88 @@ public class OrdersService : AbstractService
 
         public async Task<OrderResponse> PlaceMarketOrderAsync(
```
please remove the unused using statement at the top

dougdellolio commented on GDAXSharp/Services/AbstractService.cs:
```diff
@@ -15,6 +16,16 @@ public abstract class AbstractService
+            NullValueHandling = NullValueHandling.Ignore,
+            ContractResolver = new DefaultContractResolver
+            {
+                NamingStrategy = new SnakeCaseNamingStrategy()
```
didn't know you could do this. learning something new 😃

chrisw000 commented:
I've made those changes.

RE OrderStatus values; "Open" and "Done" were also missing... so added those.

This: 
https://stackoverflow.com/questions/48132078/available-gdax-order-statuses-and-meanings

mentions a value of "Received" - but that gives a BadRequest when searching and I've not seen it come back.

I've noticed that there is an inconsistency in where the Enums are held in the namespace (mostly due to me!). Currently their either in the .Models folder, or up one level in the .Services folder.

Happy to tidy this up - take your pick for where to put them (or put all Enums into a GDAXSharp.Enums namespace?)

Finally; I've also slipped in the #83 fix changing:   const int maxPeriods = 300;

dougdellolio commented:
@chrisw000 excellent work. merging this in now. any further changes/refactoring we can do in separate pull requests.
